react-native-community/cli
# feat: Include transitive deps in autolinking

Summary:
---------
Original issue: #870 

Currently autolinking mechanism is using only dependencies explicitly defined in `package.json`. However, when you install package dependent on other package containing native code, these additional dependencies are not autolinked and developer needs to install them separately to make them visible for autolinking mechanism.

The script is going through all the dependencies recursively, collecting them into a single list with info if they contain any peer dependencies declared. Then, it removes all the peer dependencies that are already declared in `package.json` and those peer dependencies which do not contain any native code. It displays the list of all dependencies where any peer dependency with native code is missing and asks to install them. It gets the highest possible version of the library based on given ranges declared in peer dependencies and adds the library under `dependencies` with caret range. If it can't find any version that satisfies all the ranges, it displays warning and skips this library from being installed. After installation, it asks to install pods again. After installing it, the process of running/building an app continues.

At this moment I hide it behind a `--dependency-check` flag to make sure it's optional for developers. Enabling it for everyone would block CI pipelines where any dependencies are missing. This also opens up a possibility to test it for app developers and library maintainers.

https://github.com/react-native-community/cli/assets/13985840/3c83060f-18cd-44dc-9beb-3e88aa3a3476

~~To improve that flow I created a function that goes through dependencies recursively, collects all the dependencies regardless if they are introduced in app's dependencies, and additionally looks for its duplicates with different version. If there are any duplicates, the latest version is picked for autolinking.~~

~~This way libraries can declare packages dependent on other packages containing native code in `dependencies`. No matter if the latest version of the library is on the top-level of `node_modules` or in a nested one, the script will find it and pass the correct path to the config object.~~

Test Plan:
1. Set up the repository following the contributing guide
2. Create a fresh React Native project using CLI
3. Install the following packages using yarn/npm: `@react-navigation/native @react-navigation/stack @gorhom/bottom-sheet react-navigation-tabs`. The `react-navigation-tabs` is necessary for testing incompatibility between peer dependencies versions
4. Run `yarn ios --dependency-check` / `npm run ios --dependency-check` to verify if there are any dependencies containing native code that need to be installed.
5. Verify that the list contains `@react-navigation/elements` which is not a direct dependency of an app, but it's a dependency of `@react-navigation/stack` and contains `react-native-safe-area-context` as peer dependency.
6. Confirm installing missing dependencies
7. Verify that you have been warned about incompatibility of `react-native-reanimated` version
8. Verify that all the others installed libraries are now under `dependencies` in `package.json`
9. Confirm pod installation to make sure transitive dependencies are linked properly
10. After going through all these steps, the process will continue.

The same process will also work for `build-ios`, `run-android` and `build-android` commands. You can also test it on an existing app, to see if there are any peer dependencies containing native code that are not visible for autolinking mechanism.

----------

Checklist
----------
- [x] npm support
- [x] yarn support
- [x] hide dependency check behind the flag 
- [x] write test plan
- [ ] [Native code compatibility checks](https://github.com/react-native-community/cli/pull/2054#issuecomment-1691164501)
- [ ] Write/update tests
- [ ] Test it with monorepos
- [ ] Add peer deps check to `doctor`
